File: 2080-range-frequency-queries/2080-range-frequency-queries.cpp
class RangeFreqQuery {
private:
    unordered_map<int, vector<int>> um;

public:
    RangeFreqQuery(vector<int>& arr) {
        for (int i = 0; i < (int)arr.size(); ++i) {
            um[arr[i]].push_back(i);
        }
    }
    
    int lb(vector<int> &v, int target) {
        int n = v.size();
        int l = 0, r = n - 1;
        int res = n;
        while (l <= r) {
            int m = (l + r) >> 1;
            if (v[m] < target) {
                l = m + 1;
            }
            else {
                res = m;
                r = m - 1;
            }
        }
        return res;
    }

    int ub(vector<int> &v, int target) {
        int n = v.size();
        int l = 0, r = n - 1;
        int res = n;
        while (l <= r) {
            int m = (l + r) >> 1;
            if (v[m] <= target) {
                l = m + 1;
            }
            else {
                res = m;
                r = m - 1;
            }
        }
        return res;
    }
    
    int query(int left, int right, int value) {
        vector<int> &v = um[value];
        int idx1 = ub(v, right), idx2 = lb(v, left);
        return idx1 - idx2;
    }
};

/**
 * Your RangeFreqQuery object will be instantiated and called as such:
 * RangeFreqQuery* obj = new RangeFreqQuery(arr);
 * int param_1 = obj->query(left,right,value);
 */

File: 2080-range-frequency-queries/2080-range-frequency-queries.py
from sortedcontainers import SortedList

class RangeFreqQuery:

    def __init__(self, arr: List[int]):
        self.hashMap = defaultdict(SortedList)
        n = len(arr)
        
        for i in range(n):
            self.hashMap[arr[i]].add(i)

    def query(self, left: int, right: int, value: int) -> int:
        c = self.hashMap[value]
        return c.bisect_right(right) - c.bisect_left(left)


# Your RangeFreqQuery object will be instantiated and called as such:
# obj = RangeFreqQuery(arr)
# param_1 = obj.query(left,right,value)